### PR TITLE
Feature: `Content-Type` should be set using a custom MimeDB

### LINKED ISSUE #883: `CONTENT_TYPE` header gets incorrect values because of multiple bugs with `Files.probleFileContent` used in its implementation.
**Describe the bug**
FileType detections using `Files.probeFileContent` is platform-specific and largely unreliable.
`CONTENT_TYPE` header is `null` (when serving files with extension '.mp4' ) on openJDK because Files.probleFileContent returns `null`.

File type detection is largely platform-specific and was not considered reliable.
**To Reproduce**
1. Run the example `FileStreaming.scala` on `Linux` os with `Open JDK11`
2. Request 'http://www.localhost:8090/video' throws an exception

**Expected behavior**
We should see a proper response with `CONTENT_TYPE` header set as 'video/mp4`

**Screenshots**

**Platform (please complete the following information):**
Ubuntu 11.04 + OpenJDK

**Additional context**
A few bugs being tracked with `Files.probeFileContent`

https://bugs.openjdk.java.net/browse/JDK-7142921
https://bugs.openjdk.java.net/browse/JDK-8186071
https://bugs.openjdk.java.net/browse/JDK-8129632

### PR DESCRIPTION
**Pull request by ashprakasan**
Fixes #883 

`MimeDb.scala` should ideally be auto-generated from https://www.iana.org/assignments/media-types/media-types.xhtml. As of now, the list of `MediaType` is hard-coded.

### CODE REVIEW

**Comment by tusharmath** [comment]
Take a pull from `dream11:main`

**Comment by ashprakasan** [comment]
As I understand, there are a few different options for auto-generating `MimeDB.scala`

1. Generating code using string templates ([Twirl](https://github.com/playframework/twirl)).

     **Pros** : This seems to have the easiest learning curve of all the approaches. 
                Twirl works very well with scala. The syntax is `scala-ish`
                Code is easy to read and understand.

     **Cons** : Can produce code that doesn't compile.

2. Generating code from an abstract syntax tree ([treehugger](http://eed3si9n.com/treehugger/)).

      **Pros** : Uncompilable code cannot be generated.
                        Uses scala syntax.

      **Cons** : Steep learning curve. Will take me some time to do this.
                        Not much IDE support.
                        Code looks complex, reader might take some time to understand what is happening.
                         
3. Building an abstract syntax tree and sending it directly to the compiler ([Scala macros](https://scalamacros.org/)).
      
      **Pros** : Directly passes bytecode of AST to next phase of compilation without generating `.scala` file
                       Syntactically invalid code will not be generated.

      **Cons** : Steep learning curve. Will take me some time to complete this.
                        Not much IDE support.
                        Code can be complex. Reader needs to know macros to understand what is happening.

What do you think will be a good approach here? @tusharmath, @amitksingh1490

**Comment by ashprakasan** [comment]
[#934](https://github.com/dream11/zio-http/issues/934) was created to separately address the auto-generation of `MimeDB.scala`

File: zio-http/src/main/scala/zhttp/http/MediaType.scala [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+  val fileExtensions: List[String] = Nil,
+  extensions: Map[String, String] = Map.empty,
+)
+
+object MediaType extends MimeDB {
+  def forExtention(ext: String): Option[MediaType] = extensionMap.get(ext.toLowerCase)
+
+  val extensionMap: Map[String, MediaType] = allMediaTypes.flatMap(m => m.fileExtensions.map(_ -> m)).toMap
+
+  def probeContentType(name: String): Option[String] = name.lastIndexOf(".") match {
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/http/MediaType.scala]: Let's have an unsafe implementation which is package private also. That returns a `String` or `null`.

Purely for perf reasons.

File: zio-http/src/main/scala/zhttp/http/MediaType.scala [hunk 2]
```diff
@@ -0,0 +1,21 @@
+package zhttp.http
+
+case class MediaType(
+  mainType: String,
+  val subType: String,
+  val compressible: Boolean = false,
+  val binary: Boolean = false,
+  val fileExtensions: List[String] = Nil,
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/http/MediaType.scala]: We don't need `val`. It's a `case class`

File: zio-http/src/test/scala/zhttp/http/ContentTypeSpec.scala [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+            .map(_.getOrElse("Content type header not found."))
+        assertM(res)(equalTo("application/javascript"))
+      } +
+      testM("no extension") {
+        val file = new File(getClass.getResource("/TestFile4").getPath)
+        val res  =
+          Http
+            .fromFile(file)
+            .deploy
+            .getHeaderValue(HttpHeaderNames.CONTENT_TYPE)
```
Comment (tusharmath) [inline zio-http/src/test/scala/zhttp/http/ContentTypeSpec.scala]:
```suggestion
            .getContentType
```
Everywhere.

File: zio-http/src/test/scala/zhttp/http/ContentTypeSpec.scala [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+        val file = new File(getClass.getResource("/truststore.jks").getPath)
+        val res  =
+          Http
+            .fromFile(file)
+            .deploy
+            .getHeaderValue(HttpHeaderNames.CONTENT_TYPE)
+            .run()
+            .map(_.getOrElse("Content type header not found."))
+        assertM(res)(equalTo("Content type header not found."))
+      }
```
Comment (tusharmath) [inline zio-http/src/test/scala/zhttp/http/ContentTypeSpec.scala:78]: I think all the tests are doing the same thing. We can maybe reduce the boilerplate.
It's a nice idea to test for multiple extensions.

Comment (ashprakasan) [reply zio-http/src/test/scala/zhttp/http/ContentTypeSpec.scala:78]: okay.

File: zio-http/src/test/scala/zhttp/http/ContentTypeSpec.scala (hunk 3 above)
Comment (ashprakasan) [reply zio-http/src/test/scala/zhttp/http/ContentTypeSpec.scala]: done.

File: zio-http/src/main/scala/zhttp/http/MediaType.scala (hunk 2 above)
Comment (ashprakasan) [reply zio-http/src/main/scala/zhttp/http/MediaType.scala]: Done.

File: zio-http/src/main/scala/zhttp/http/MediaType.scala (hunk 1 above)
Comment (ashprakasan) [reply zio-http/src/main/scala/zhttp/http/MediaType.scala]: I've memoised the result for now, will do other perf enhancements in the later PRs.

File: zio-http/src/main/scala/zhttp/http/MediaType.scala [hunk 5]
```diff
@@ -0,0 +1,34 @@
+package zhttp.http
+
+import java.util
+
+case class MediaType(
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/http/MediaType.scala]:
```suggestion
final class MediaType private[zhttp](
```

The MediaType should only be constructed in `MimeDB`

File: zio-http/src/main/scala/zhttp/http/Http.scala [hunk 6]
```diff
@@ -203,6 +204,13 @@ sealed trait Http[-R, +E, -A, +B] extends (A => ZIO[R, Option[E], B]) { self =>
   final def headerValue(name: CharSequence)(implicit eb: IsResponse[B]): Http[R, E, A, Option[CharSequence]] =
     headers.map(_.headerValue(name))
 
+  /**
+   * Extracts the value of ContentType header
+   */
+  final def contentType(implicit eb: IsResponse[B]): Http[R, E, A, Option[CharSequence]] = headerValue(
+    HttpHeaderNames.CONTENT_TYPE,
+  )
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/http/Http.scala:212]: This is a nice addition!

File: zio-http/src/main/scala/zhttp/http/MimeDB.scala [hunk 7]
```diff
@@ -0,0 +1,7787 @@
+package zhttp.http
+
+private[zhttp] trait MimeDB {
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/http/MimeDB.scala:3]: We should be able to set and get content type via MimeDB in `Headers`, `Request` and `Response`

File: zio-http/src/main/scala/zhttp/http/Http.scala (hunk 6 above)
Comment (tusharmath) [reply zio-http/src/main/scala/zhttp/http/Http.scala:212]: Should this return a `MediaType`?

File: zio-http/src/main/scala/zhttp/http/MediaType.scala (hunk 5 above)
Comment (ashprakasan) [reply zio-http/src/main/scala/zhttp/http/MediaType.scala:5]: This can be used to add new `MediaType`

**Review by tusharmath — APPROVED** [review]

**Pull request merged**